activerecord/lib/active_record/type/decimal_without_scale.rb

```diff
@@ -10,6 +10,13 @@ def type
       def type_cast_for_schema(value)
         value.to_s.inspect
       end
+
+      private
+
+        def cast_value(value)
+          raise ActiveModel::RangeError, "cannot be infinite" if value.to_d.infinite?
```

Comment:
I also don't feel good about this one. Now any value will have the overhead of this conversion even that it is going to pass this check. Can't we compare with the string representation? But I'm not sure if that comparing is faster. I also would only do it for the case that we were going to call `to_i`. There is not reason the other conditionals do this.

Reply:
`cast_value` is called when the attribute is assigned, which is too early. The error should only be raised when the record is saved, so that e.g. validations have a chance to reject the value cleanly. Moving the check to `serialize` will achieve that, and also address the performance concern: the value passed to `serialize` will always be a BigDecimal (or nil), so we can call `infinite?` on it directly.

See https://github.com/rails/rails/commit/7c6f3938dee47f0932c2a1d4924adaebc25517ac, which moved the integer range check from `cast_value` to `type_cast_for_database` (which was later renamed to `serialize`).

Reply:
> `cast_value` is called when the attribute is assigned, which is too early.

I'm not seeing that behaviour in my tests. I have to call `#save` to make them pass. I'll try moving the raise to `#serialize` regardless.

Reply:
It looks like the assigned value is type cast lazily since https://github.com/rails/rails/commit/8e633e505880755e7e366ccec2210bbe2b5436e7. Reading the attribute back still type casts it, though:

```patch
diff --git a/activerecord/test/cases/type/decimal_test.rb b/activerecord/test/cases/type/decimal_test.rb
index 4490e8837b..69a0a5a29d 100644
--- a/activerecord/test/cases/type/decimal_test.rb
+++ b/activerecord/test/cases/type/decimal_test.rb
@@ -35,6 +35,19 @@ class DecimalTest < ActiveRecord::TestCase

         assert_equal "cannot be infinite", error.message
       end
+
+      test "infinte values can be read back" do
+        klass = Class.new(ActiveRecord::Base) do
+          self.table_name = "posts"
+          attribute :foo, :decimal
+        end
+        model = klass.new
+
+        assert_nothing_raised do
+          model.foo = BigDecimal("-Infinity")
+          model.foo
+        end
+      end
     end
   end
 end
```

Reply:
I was able to move casting to `#cast_type` in `DecimalWithoutScale` and moved the raising to `#serialize` in both classes. I now also have test coverage for `DecimalWithoutScale`.